### PR TITLE
Error reporting improvements

### PR DESCRIPTION
**Pull request by Zlata-Zueva**
* Replaces all `SystemException`s with more meaningful errors and improves messages.
* Makes sure exceptions get stringified properly.
* Adds error logs to tests on failures during resources clean up.

Partly addresses #12, but there are still places to improve.

TODO:
 - [x] Update changelog
 - [x] Bump up SDK version

### CODE REVIEW

**Comment by larf311** [comment]
I notice the Java SDK just returns the HTTP errors directly (e.g. 404 for not found). I suspect that was a deliberate choice by @bradlo and that we should probably do the same here.

File: RelationalAI/Errors/ApiException.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        public ApiException(
+            string message,
+            HttpStatusCode statusCode,
+            string response,
+            HttpResponseHeaders headers)
+            : base($"Request failed with error: {message}")
+        {
+            StatusCode = statusCode;
+            Response = response;
+            Headers = headers;
```
Comment (bradlo) [inline RelationalAI/Errors/ApiException.cs:37]: am curious why the api exception carries the HTTP headers? in the other SDKs we generally try to abstract away transport details at the API layer.

**Review by bradlo — APPROVED** [review]
Looks good to me. One question, but mostly just curious. And thank you for the updates to the changelog.

File: RelationalAI/Errors/ApiException.cs (hunk 1 above)
Comment (stanislav-bedunkevich) [reply RelationalAI/Errors/ApiException.cs:37]: @bradlo we really need that for the request id, as that's always important for tracing. We could extract the X-Request-ID header value and pass it directly if we know the exact header name.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## v0.8.0-alpha
+* Added the following exceptions:
+  - `NotFoundException` thrown from `GetEngineAsync(string engine)`, `GetDatabaseAsync(string database)`, etc. when requested resource (Engine/Database/Model/User/Client) doesn't exist or got deleted.
+  - `ApiException` thrown when RAI API responds with 5xx status codes or contains unsupported content type.
+  - `EngineProvisionFailedException` thrown from `CreateEngineWaitAsync` when requested engine failed to provision.
+  - `CredentialsNotSupportedException` thrown from every method when credentials to access RAI API are not provided or provided credentials are unsupported (i.e. not for OAuth Client credentials method).
+  - `InvalidResponseExcetion` thrown when RAI API response has unexpected format or content type.
```
Comment (stanislav-bedunkevich) [inline CHANGELOG.md:9]:
```suggestion
  - `InvalidResponseException` thrown when RAI API response has unexpected format or content type.
```

**Review by stanislav-bedunkevich — APPROVED** [review]

**Review by larf311 — APPROVED** [review]

File: RelationalAI/Services/Rest.cs [hunk 3]
```diff
@@ -265,11 +268,29 @@ private static List<TransactionAsyncFile> ParseMultipartResponse(byte[] content)
             return output;
         }
 
+        private static async Task EnsureSuccessResponseAsync(HttpResponseMessage response)
+        {
+            var status = (int)response.StatusCode;
+            if (status != 404 && status < 500)
+            {
+                return;
+            }
```
Comment (NRHelmi) [inline RelationalAI/Services/Rest.cs:277]: what happens if resource if forbidden (403) ? Do we swallow client issues other than 404 ?

File: RelationalAI/Services/Rest.cs [hunk 4]
```diff
@@ -293,14 +314,14 @@ private async Task<AccessToken> RequestAccessTokenAsync(string host, ClientCrede
             var resp = await RequestHelperAsync("POST", creds.ClientCredentialsUrl, data);
             if (!(resp is string stringResponse))
             {
-                throw new SystemException("Unexpected response type");
+                throw new InvalidResponseException("Unexpected response type, expected a string", resp);
```
Comment (NRHelmi) [inline RelationalAI/Services/Rest.cs:317]: maybe also include the current response type in the exception message

Comment (Zlata-Zueva) [reply RelationalAI/Services/Rest.cs]: @NRHelmi addressed in the latest commit. Thank you.

File: RelationalAI/Services/Rest.cs (hunk 3 above)
Comment (stanislav-bedunkevich) [reply RelationalAI/Services/Rest.cs:277]: @NRHelmi yeah, we mostly wanted to leave the behaviour as it was before. I think it's better for the SDK team to agree on the http status code handling to be consistent across all of the SDKs.

The reason for 5xx is simple, as we run into them (occasionally we do), we want to be able to see the request id in logs.

Your point is very valid though, but we'd like to leave it out of the scope of this PR is that's fine with you?

Comment (NRHelmi) [reply RelationalAI/Services/Rest.cs:277]: Thank you for the clarification, can we create an issue as part of the project ?
cc @bradlo

**Review by NRHelmi — APPROVED** [review]
thank you :)

File: RelationalAI/Services/Rest.cs (hunk 4 above)
Comment (NRHelmi) [reply RelationalAI/Services/Rest.cs]: thank you :pray:

**Pull request merged**